File: templates/Enterprise-Template/src/typescript/src/dialogs/authentication/authenticationDialog.ts
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License

import { User } from '@microsoft/microsoft-graph-types';
import { TokenResponse, TurnContext } from 'botbuilder';
import { ComponentDialog, DialogTurnResult, OAuthPrompt, WaterfallDialog, WaterfallStepContext } from 'botbuilder-dialogs';
import { GraphClient } from '../../serviceClients/graphClient';
import { AuthenticationResponses } from './authenticationResponses';

export class AuthenticationDialog extends ComponentDialog {
    private static readonly _responder: AuthenticationResponses = new AuthenticationResponses() ;

    // Fields
    private _connectionName: string;

    constructor(connectionName: string) {
        super(AuthenticationDialog.name);
        this.initialDialogId = AuthenticationDialog.name;
        this._connectionName = connectionName;

        const authenticate = [
            this.prompToLogin.bind(this),
            this.finishLoginhDialog.bind(this)
        ];

        this.addDialog(new WaterfallDialog(this.initialDialogId, authenticate));
        this.addDialog(new OAuthPrompt(DialogIds.LoginPrompt, {
            connectionName: this._connectionName,
            text: i18n.__('authentication.prompt'),
            title: i18n.__('authentication.title')
        }));
    }

    private prompToLogin(sc: WaterfallStepContext): Promise<DialogTurnResult> {
        return sc.prompt(AuthenticationResponses.ResponseIds.LoginPrompt, {});
    }

    private async finishLoginhDialog(sc: WaterfallStepContext): Promise<DialogTurnResult> {

        if (sc.result) {
            const tokenResponse: TokenResponse = sc.result;

            if (tokenResponse.token) {
                const user = await this.getProfile(sc.context, tokenResponse);
                await AuthenticationDialog._responder.replyWith(sc.context, AuthenticationResponses.ResponseIds.SucceededMessage, { name: user.displayName });
                return sc.endDialog(tokenResponse);
            }

        } else {
            await AuthenticationDialog._responder.replyWith(sc.context, AuthenticationResponses.ResponseIds.FailedMessage);
        }

        return sc.endDialog();
    }

    private getProfile(context: TurnContext, tokenResponse: TokenResponse): Promise<User> {
        const token = tokenResponse;
        const client: GraphClient = new GraphClient(token.token);

        return client.getMe();
    }
}

class DialogIds {
    public static LoginPrompt : string =  'loginPrompt';
}

File: templates/Enterprise-Template/src/csharp/EnterpriseBotSample/EnterpriseBotSample/Dialogs/Authentication/AuthenticationDialog.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

using System.Threading;
using System.Threading.Tasks;
using EnterpriseBotSample.Dialogs.Authentication.Resources;
using EnterpriseBotSample.ServiceClients;
using Microsoft.Bot.Builder;
using Microsoft.Bot.Builder.Dialogs;
using Microsoft.Bot.Schema;
using Microsoft.Graph;

namespace EnterpriseBotSample.Dialogs.Shared
{
    public class AuthenticationDialog : ComponentDialog
    {
        private static AuthenticationResponses _responder = new AuthenticationResponses();

        public AuthenticationDialog(string connectionName)
            : base(nameof(AuthenticationDialog))
        {
            InitialDialogId = nameof(AuthenticationDialog);
            ConnectionName = connectionName;

            var authenticate = new WaterfallStep[]
            {
                PromptToLogin,
                FinishLoginDialog,
            };

            AddDialog(new WaterfallDialog(InitialDialogId, authenticate));
            AddDialog(new OAuthPrompt(DialogIds.LoginPrompt, new OAuthPromptSettings()
            {
                ConnectionName = ConnectionName,
                Title = AuthenticationStrings.TITLE,
                Text = AuthenticationStrings.PROMPT,
            }));
        }

        private string ConnectionName { get; set; }

        private async Task<DialogTurnResult> PromptToLogin(WaterfallStepContext sc, CancellationToken cancellationToken)
        {
            return await sc.PromptAsync(AuthenticationResponses.ResponseIds.LoginPrompt, new PromptOptions());
        }

        private async Task<DialogTurnResult> FinishLoginDialog(WaterfallStepContext sc, CancellationToken cancellationToken)
        {
            var activity = sc.Context.Activity;
            if (sc.Result != null)
            {
                var tokenResponse = sc.Result as TokenResponse;

                if (tokenResponse?.Token != null)
                {
                    var user = await GetProfile(sc.Context, tokenResponse);
                    await _responder.ReplyWith(sc.Context, AuthenticationResponses.ResponseIds.SucceededMessage, new { name = user.DisplayName });
                    return await sc.EndDialogAsync(tokenResponse);
                }
            }
            else
            {
                await _responder.ReplyWith(sc.Context, AuthenticationResponses.ResponseIds.FailedMessage);
            }

            return await sc.EndDialogAsync();
        }

        private async Task<User> GetProfile(ITurnContext context, TokenResponse tokenResponse)
        {
            var token = tokenResponse;
            var client = new GraphClient(token.Token);

            return await client.GetMe();
        }

        private class DialogIds
        {
            public const string LoginPrompt = "loginPrompt";
        }
    }
}
